### PR TITLE
Specification of the light client

### PR DESCRIPTION
**Pull request by konnov**
The TLA+ specification of the light client, including bisection. TLC checked this one for the height 3.

### CODE REVIEW

File: spec/light-client/Lightclient.tla [hunk 1]
```diff
@@ -1,111 +1,273 @@
...
+            signedHdr[2] \intersect DOMAIN trustedHdr.NextVP)
+      IN
+      IF heightToVerify = heightToTrust + 1
+          \* the special case of adjacent heights: check 2/3s
+      THEN (3 * SP > 2 * TP)
+      ELSE \* the general case: check 1/3 between the headers and make sure there are 2/3 votes  
+        LET TPV == BC!PowerOfSet(signedHdr[1].VP, DOMAIN signedHdr[1].VP)
+            SPV == BC!PowerOfSet(signedHdr[1].VP, signedHdr[2])
+        IN
+        (3 * SP > TP) /\ (3 * SPV > 2 * TPV)
```
Comment (ancazamfir) [inline spec/light-client/Lightclient.tla]: I am not sure if it is a good idea to group all checks in this function. We can have 3 types of failures here:
1 - outside trusting period
2 - 2/3rds have not signed
3 - less than 1/3rd common validators between trusted and signed

I believe bisection should be invoked as long as 1 or 3 happen. If neither happens we should check 2 and if true declare failure and go to `finished`. 
It is possible also that we do 2 first and then 1+3. But the idea is that 2 should cause a hard stop...unless we want to model connecting to multiple peers which I think it's not in scope.

File: spec/light-client/Lightclient.tla [hunk 2]
```diff
@@ -1,111 +1,273 @@
...
+             Sequential verification tells us that the header cannot be trusted:
+             If the search request was scheduled as a left branch, then
+             (1) pop the top request (the left branch), and
+             (2) pop the second top request, which is the request for the right branch.
+             Otherwise, pop only the top request.
+             This is the optimization that is introduced in the English spec.
+             One can replace it e.g. with a parallel search.
+             *)
+            LET howManyToPop == IF topReq.isLeft THEN 2 ELSE 1 IN  
+            <<FALSE, SubSeq(requestStack, 1 + howManyToPop, Len(requestStack))>>
```
Comment (ancazamfir) [inline spec/light-client/Lightclient.tla]: First I think an explanation on what SubSeq would be useful:
`SubSeq(s, m, n) == [i \in 1..(1+n-m) |-> s[i+m-1]]
  (* The sequence <<s[m], s[m+1], ... , s[n]>>.                            *)
`
If we start with an `initStack` like: `<<(T, N, N+1)>>` and we are at 181 with failed `CheckSupport` then:
- 181: `howManyToPop` == 2
- 182: will call `SubSeq(initStack, 3, 1)` -> I assume this will just return empty sequence. Just a bit awkward.

But another potential problem is that in 182 we may have non-empty stack which takes the caller to line 206 where the verdict is not checked and we continue verification of right branch with a failure in left. In general I don't understand this optimization. Which english spec you refer to here?

File: spec/light-client/Lightclient.tla [hunk 3]
```diff
@@ -1,111 +1,273 @@
...
+    (* an index of the block header that the light client trusts by social consensus *)
+  TO_VERIFY_HEIGHT
+    (* an index of the block header that the light client tries to verify *)
+
+VARIABLES
+  state,        (* the current state of the light client *)
+  inEvent,      (* an input event to the light client, e.g., a header from a full node *)
+  outEvent,     (* an output event from the light client, e.g., finished with a verdict *)
+  requestStack,     (* the stack of requests to be issued by the light client to a full node *)
+  storedHeaders     (* the set of headers obtained from a full node*)
```
Comment (ancazamfir) [inline spec/light-client/Lightclient.tla]: I feel we need some description of at least the elements of requestStack and storedHeaders, here or elsewhere. e.g. the request in the `requestStack` has `isLeft`, `startHeight`, `endHeight`. Headers have the block and the validator set. Also the `storedHeaders` behave like a set, there cannot be two headers with same height.

File: spec/light-client/Lightclient.tla (hunk 2 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: We used the [spec](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md) as the input. Essentially, when one pops two elements, the top-level Bisection call returned false and thus the inner call to Bisection should never occur. This looks tricky, as we have to encode two non-tail recursive calls using the stack (Bisection in [English spec](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md) cannot be easily converted into a loop).

File: spec/light-client/Lightclient.tla (hunk 3 above)
Comment (milosevic) [reply spec/light-client/Lightclient.tla]: It seems that Anca is also asking for types :) We need to figure out convincing way to "simulate" types as 1) that is common way engineers think about specs and 2) it is hard to follow spec without knowing what variables refer to.

File: spec/light-client/Lightclient.tla [hunk 4]
```diff
@@ -1,111 +1,273 @@
...
+VARIABLES tooManyFaults, height, minTrustedHeight, blockchain, Faulty
+
+\* All the variables of Blockchain. For some reason, BC!vars does not work
+bcvars == <<tooManyFaults, height, minTrustedHeight, blockchain, Faulty>>
+
+(* Create an instance of Blockchain.
+   We could write EXTENDS Blockchain, but then all the constants and state variables
+   would be hidden inside the Blockchain module.
+ *) 
+BC == INSTANCE Blockchain WITH tooManyFaults <- tooManyFaults, height <- height,
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: Why we need to do this?

File: spec/light-client/Lightclient.tla [hunk 5]
```diff
@@ -1,111 +1,273 @@
...
+(* The events that can be generated by the environment (reactor?):
+    user requests and node responses *)
+EnvNext ==
+    \/ /\ state = "init"
+       \* modeling feature, do not start the client before the blockchain is constructed
+       /\ height >= TO_VERIFY_HEIGHT
+       /\ inEvent' = [type |-> "start", heightToVerify |-> TO_VERIFY_HEIGHT]
+    \/ /\ state = "working"
+       /\ outEvent.type = "requestHeader"
+       /\ \E hdr \in BC!SoundSignedHeaders(outEvent.height):
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: We need more comments here. I am lost :)

File: spec/light-client/Lightclient.tla [hunk 6]
```diff
@@ -1,111 +1,273 @@
...
+            inEvent' = [type |-> "responseHeader", hdr |-> hdr]
+
+(************************** Lite client ************************************)
+
+(* the control states of the lite client *) 
+States == { "init", "working", "finished" }
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: We can probably reorder some definitions so it's a bit easier to understand what is the protocol being modelled, what is environment (system model somehow) and what are the properties we want to check.

File: spec/light-client/Lightclient.tla [hunk 7]
```diff
@@ -1,111 +1,273 @@
...
+
+(* Produce a request event for the top element of the requestStack *)
+RequestHeaderForTopRequest(stack) ==
+    IF stack = <<>>
+    THEN NoEvent
+    ELSE  LET top == Head(stack)
+              heightToRequest ==
+                IF top.isLeft
+                THEN top.endHeight      \* the pivot is on the right
+                ELSE top.startHeight    \* the pivot is on the left
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: What is this?

File: spec/light-client/Lightclient.tla [hunk 8]
```diff
@@ -1,111 +1,273 @@
...
           IN
-          blockchain = <<genesis>>
+          [type |-> "requestHeader", height |-> heightToRequest]
+
+(* When starting the light client *)
+OnStart ==
+    /\ state = "init"
+    /\ inEvent.type = "start"
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: This is confusing.

File: spec/light-client/Lightclient.tla (hunk 2 above)
Comment (ancazamfir) [reply spec/light-client/Lightclient.tla]: Yes but the bisection should stop only if >2/3rds has not signed. In the other cases it should continue.

File: spec/light-client/Lightclient.tla (hunk 3 above)
Comment (ancazamfir) [reply spec/light-client/Lightclient.tla]: At a minimum we should explain in the comments.

Comment (milosevic) [reply spec/light-client/Lightclient.tla]: Short term we need to figure out how to take maximum out of TLA+. Mid/Long term I think we should explore alternatives, for example something like Stainless. As main users of this spec are engineers, we probably need something closer to the languages/tools they get used to.

File: spec/light-client/Lightclient.tla [hunk 9]
```diff
@@ -1,111 +1,273 @@
...
+            SPV == BC!PowerOfSet(signedHdr[1].VP, signedHdr[2])
+        IN
+        (3 * SP > TP) /\ (3 * SPV > 2 * TPV)
+
+(* Make one step of bisection, roughly one stack frame of Bisection in the English spec *)
+OneStepOfBisection(storedHdrs) ==
+    LET topReq == Head(requestStack)
+        lh == topReq.startHeight
+        rh == topReq.endHeight
+        lhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = lh
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: Would be probably better using records instead of tuples if possible for readability purpose.

File: spec/light-client/Lightclient.tla [hunk 10]
```diff
@@ -1,111 +1,273 @@
...
+             *)
+            LET rightReq == [isLeft |-> FALSE, startHeight |-> (lh + rh) \div 2, endHeight |-> rh]
+                leftReq ==  [isLeft |-> TRUE, startHeight |-> lh, endHeight |-> (lh + rh) \div 2]
+            IN
+            <<TRUE, <<leftReq, rightReq>> \o Tail(requestStack)>>
+
+OnResponseHeader ==
+  /\ state = "working"
+  /\ inEvent.type = "responseHeader"
+  /\ storedHeaders' = storedHeaders \union { inEvent.hdr }
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: Should we store all headers or only valid one?

File: spec/light-client/Lightclient.tla (hunk 3 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: The standard practice in TLA+ is to write an invariant `TypeOK`, which basically checks variable types in every state. I will add one like that.

Moving to Stainless or similar tools will not solve all our problems, it will only give better assurance that the code matches the specifications. But this will come with the price of investing more efforts in writing pre-, post-conditions, and the invariants, as well as making them amenable for verification.

File: spec/light-client/Lightclient.tla [hunk 11]
```diff
@@ -1,111 +1,273 @@
...
+\* Correctness states that all the obtained headers are exactly like in the blockchain.
+\* This formula is equivalent to Accuracy in the English spec, that is, A => B iff ~A => ~B.
+CorrectnessInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = TRUE
+        => (\A hdr \in storedHeaders: hdr[1] = blockchain[hdr[1].height])
+
+PrecisionInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = FALSE
+        => (\E hdr \in storedHeaders: hdr[1] /= blockchain[hdr[1].height])
+
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: Can we define a property that light client (wrongly) trusted a header that is different from the one in the blockchain?

Comment (milosevic) [reply spec/light-client/Lightclient.tla]: We might want to modify `Lite Client Accuracy` property to never trust a header which is outside trusted period. More general, it would be great if these specs are standalone, i.e., that reader does not need to read English spec. Probably English spec should appear in form of comments here.

File: spec/light-client/Lightclient.tla [hunk 12]
```diff
@@ -1,111 +1,273 @@
...
+    outEvent.type = "finished" /\ outEvent.verdict = FALSE
+        => (\E hdr \in storedHeaders: hdr[1] /= blockchain[hdr[1].height])
+        
+\* TODO: add Termination that says that the light client always reaches the finished state
+\* under some conditions (when communicating with a correct full node).
+
+\* TODO: specify Completeness, which is a combination of Termination and Precision.
+\* TODO: Completeness in the English spec assumes that the lite client communicates
+\* with a correct full node.  
+
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: We should probably also add equivalent of `Verification Condition` (English spec) property.

File: spec/light-client/Lightclient.tla (hunk 1 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: Fixed after our discussions, see b06216bb54f8d26a8b1d6cc0123546dcbcb18957

File: spec/light-client/Lightclient.tla (hunk 2 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: Fixed. Now it is much simpler. See b06216bb54f8d26a8b1d6cc0123546dcbcb18957

File: spec/light-client/Lightclient.tla (hunk 3 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: See `TypeOK` in b06216bb54f8d26a8b1d6cc0123546dcbcb18957

File: spec/light-client/Lightclient.tla (hunk 4 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: An alternative would be to write `EXTENDS Blockchain` to include in the spec. But then Blockchain's parameters would be implicit, which I do not like.

File: spec/light-client/Lightclient.tla (hunk 5 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: See e0804aceff691e248fb9aea4a70017d74704bd0a

File: spec/light-client/Lightclient.tla [hunk 13]
```diff
@@ -1,111 +1,273 @@
...
+        => (\A hdr \in storedHeaders: hdr[1] = blockchain[hdr[1].height])
+
+PrecisionInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = FALSE
+        => (\E hdr \in storedHeaders: hdr[1] /= blockchain[hdr[1].height])
+        
+\* TODO: add Termination that says that the light client always reaches the finished state
+\* under some conditions (when communicating with a correct full node).
+
+\* TODO: specify Completeness, which is a combination of Termination and Precision.
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: How we can check the case when during a bisection algorithm trusted header expires?

File: spec/light-client/Lightclient.tla [hunk 14]
```diff
@@ -1,111 +1,273 @@
...
+   This operator is similar to CheckSupport in the English spec.
+   
+   The parameters have the following meanings:
+   - heightToTrust is the height of the trusted header
+   - heightToVerify is the height of the header to be verified
+   - trustedHdr is the trusted header (not a signed header)
+   - signedHdr is the signed header to verify (including commits)
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
+    IF minTrustedHeight > heightToTrust \* outside of the trusting period
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: What if we are just at the border line (in terms of trusted period) and this condition evaluates to false, but then during the execution of this function the trusted header leaves trusted period? Is there a way that we add a check that might catch this (potential) bug?

File: spec/light-client/Lightclient.tla [hunk 15]
```diff
@@ -1,111 +1,273 @@
...
+   following the trusting period method.
+   This operator is similar to CheckSupport in the English spec.
+   
+   The parameters have the following meanings:
+   - heightToTrust is the height of the trusted header
+   - heightToVerify is the height of the header to be verified
+   - trustedHdr is the trusted header (not a signed header)
+   - signedHdr is the signed header to verify (including commits)
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: I am wondering if `CheckSupport` should return a value (instead of `boolean`) that would correspond to lease duration, i.e., for how long I can still trust this header?

File: spec/light-client/Lightclient.tla (hunk 14 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: Why is it a bug? We should understand expectations on time constraints.

File: spec/light-client/Lightclient.tla (hunk 10 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: All, as we can use them later as a proof.

File: spec/light-client/Lightclient.tla [hunk 16]
```diff
@@ -128,33 +129,47 @@ OnStart ==
...
       LET TP == BC!PowerOfSet(trustedHdr.NextVP, DOMAIN trustedHdr.NextVP)
           SP == BC!PowerOfSet(trustedHdr.NextVP,
             signedHdr[2] \intersect DOMAIN trustedHdr.NextVP)
       IN
-      IF heightToVerify = heightToTrust + 1
-          \* the special case of adjacent heights: check 2/3s
-      THEN (3 * SP > 2 * TP)
-      ELSE \* the general case: check 1/3 between the headers and make sure there are 2/3 votes  
+      IF heightToVerify = heightToTrust + 1 \* adjacent headers
+      THEN signedHdr[1].VP = trustedHdr.NextVP
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: In case of consecutive headers we still want to check if signedHdr is valid, i.e., that there are more than 2/3 of voting power signed.

File: spec/light-client/Lightclient.tla [hunk 17]
```diff
@@ -164,66 +179,83 @@ OneStepOfBisection(storedHdrs) ==
...
+        THEN (* Sequential verification tells us that the header cannot be trusted. *)
+            <<FALSE, << (* empty stack *)>> >> \* TERMINATE immediately
         ELSE (*
              Dichotomy: schedule search requests for the left and right branches
              (and pop the top element off the stack).
              In contrast to the English spec, these requests are not processed immediately,
-             but one-by-one in a depth-first order (mind the optimization above).
+             but one-by-one in a depth-first order.
              *)
             LET rightReq == [isLeft |-> FALSE, startHeight |-> (lh + rh) \div 2, endHeight |-> rh]
```
Comment (milosevic) [inline spec/light-client/Lightclient.tla]: It might be simpler expressing this logic as a state machine where initial input is end header height to be verified and then state machine have only one pending request at a time and it is trying to verify end header by executing bisection logic. State machine can probably just maintain current header and header to verify (initially half of end header and initial trusted header) and then we update those values upon successful header verification.

Comment (konnov) [reply spec/light-client/Lightclient.tla]: I believe that the [English spec](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md) is explicitly against this approach, as the inner call to Bisection is done only if the outer call to Bisection has succeeded. It does not look like we can do that without a stack.

File: spec/light-client/Lightclient.tla (hunk 11 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: `PositiveBeforeTrustedHeaderExpires` requires that whenever the lite client accepts the header, the trusted header is still within the trusting period. So far, TLC has not found a counterexample, but I believe it should find one.

**Comment by cwgoes** [comment]
Is it possible to alter permissions on this repository so that I can self-request a review?

At present, I don't seem to be able to do so.

Looking forward to the TLA+ spec!

**Comment by konnov** [comment]
Hi Christopher, I am not sure how to do it at github. You should be on the list of collaborators now.

File: spec/light-client/Lightclient.tla (hunk 9 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: done

File: spec/light-client/Lightclient.tla (hunk 13 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: The property `PositiveBeforeTrustedHeaderExpires` states that whenever the lite client finished with a positive verdict, `minTrustedHeight` is not larger than `TRUSTED_HEIGHT`, meaning that the trusted header has not expired yet. Most likely, this property does not hold true, but TLC has not found a bug yet.

File: spec/light-client/Lightclient.tla (hunk 15 above)
Comment (konnov) [reply spec/light-client/Lightclient.tla]: The lite client could return three kinds of values: red (cannot trust), yellow (can trust, but the headers expire soon), green (can trust).

**Comment by konnov** [comment]
A question for discussion: when the lite client starts, should it be the case that `(now - inithead.time) <= trustingPeriod`. The TLA+ spec is assuming that if this is not true, the lite client should not trust the initial header and the header to be verified.

**Comment by milosevic** [comment]
> A question for discussion: when the lite client starts, should it be the case that `(now - inithead.time) <= trustingPeriod`. The TLA+ spec is assuming that if this is not true, the lite client should not trust the initial header and the header to be verified.

The invariant above should always hold (also upon initialisation). It is probably simpler expressing it as `inithead.time + trustedPeriod < now`. Intuition is that trusted period of the header hasn’t expired yet.

**Comment by konnov** [comment]
I have fixed the spec after the discussion at dev-session. Additionally, the properties `Correctness` and `Precision` are refined now. All expected properties hold true on the minimal model: 2 validators + 1 full node, blockchain of length 3. This is apparently as much as we can get from TLC in a non-cloud mode.

**Comment by konnov** [comment]
I am going to merge this PR as lite client v1, so we can start: (1) verification efforts with Apalache and cloud TLC, and (2) work on v2 that does not use recursive bisection.

**Comment by konnov** [comment]
Actually, TLC just reported on a bug for `NoDupsInv`. In the current spec, the lite client makes two requests per pivot, and the full node may respond with two different headers for the same height. I will fix the spec, so the lite client would request for a pivot only once. The recursive algorithm is complicated...

**Comment by konnov** [comment]
I hope that this is the last fix for this version. TLC has found a violation of `StoredHeadersAreSoundOrNonTrusted`, as we were checking the left header instead of the right one. (Another fix is about the refinement of the property).

```diff
+++ b/spec/light-client/Lightclient.tla
@@ -195,7 +195,7 @@ OneStepOfBisection(pStoredSignedHeaders) ==
         lhdr == CHOOSE shdr \in pStoredSignedHeaders: shdr.header.height = lh
         rhdr == CHOOSE shdr \in pStoredSignedHeaders: shdr.header.height = rh
     IN
-    IF Verify(lhdr.header.VP, lhdr) = FALSE
+    IF Verify(rhdr.header.VP, rhdr) = FALSE
     THEN [verdict |-> FALSE, newStack |-> <<(* empty *)>> ] \* TERMINATE immediately
     ELSE \* pass only the header lhdr.header and signed header rhdr
       IF CheckSupport(lh, rh, lhdr.header, rhdr)
@@ -349,7 +349,7 @@ StoredHeadersAreSoundOrNotTrusted ==
     outEvent.type = "finished" /\ outEvent.verdict = TRUE
         =>
         \A left, right \in storedSignedHeaders: \* for every pair of different stored headers
-            \/ left = right
+            \/ left.header.height >= right.header.height
                \* either there is a header between them
             \/ \E middle \in storedSignedHeaders:
                 /\ left.header.height < middle.header.height
```

**Comment by konnov** [comment]
Termination, PrecisionInv, Correctness, NoDupsInv, StoredHeadersAreSoundOrNonTrusted holds true for the case of two nodes. Now we can continue with more elaborate verification.
